pkg/controller/direct/dataflow/dataflowflextemplatejob_controller.go

```diff
@@ -363,7 +390,31 @@ func (a *dataflowFlexTemplateJobAdapter) Update(ctx context.Context, updateOp *d
 	log := klog.FromContext(ctx)
 	log.V(0).Info("updating object", "u", u)
 
-	if true {
+	observedGeneration, _, err := unstructured.NestedInt64(u.Object, "status", "observedGeneration")
+	if err != nil {
+		return fmt.Errorf("reading status.observedGeneration: %w", err)
+	}
+	metadataGeneration := u.GetGeneration()
+
+	// We rely on the FlexJob being immutable, so drift at the GCP level should not be possible.
+	// Instead, we reconcile whenever we see a different spec.generation
+	if observedGeneration == metadataGeneration {
+		log.V(0).Info("object status.observedGeneration matches metadata.generation, skipping reconcile", "status.observedGeneration", observedGeneration, "metadata.generation", metadataGeneration)
+
+		j, _ := u.MarshalJSON()
+		log.V(0).Info("object is", "json", string(j))
+
+		// If we are waiting on the existing job, update the status
+		if a.actual != nil {
+			if err := a.updateStatus(ctx, updateOp, a.actual); err != nil {
+				return err
+			}
+
+			// TODO: If job fails to update, we probably need to "put back" the old job id
```

Comment:
May I ask what does "put back the old job ID" means? Does the JobID changes in pending status until it is Running? If so, it seems that we also need to handle  the `spec.resourceID`  in the case that a job is pending, and users uses a unstable JobID trying to acquire/manage that Job.

Reply:
There's a separate discussion about what the primary key for this job really should be.

Today jobs are technically immutable.  We can replace a job, but the new job gets a new job ID.  If that new job fails, the old job continues running/resumes.  In this code we now keep the new job ID, so that we don't keep launching new (failing) jobs forever.  However, it does mean that the job is still running, just with the old job ID.